src/main/java/sparksoniq/spark/iterator/flowr/ReturnClauseSparkIterator.java

```diff
@@ -49,8 +61,66 @@ public ReturnClauseSparkIterator(RuntimeTupleIterator child, RuntimeIterator exp
         return itemRDD;
     }
 
-    private JavaRDD<Item> itemRDD;
-    
-    private RuntimeTupleIterator _child;
+    @Override
+    protected boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    protected Item nextLocal() {
+        if(_hasNext == true){
+            Item result = _nextLocalResult;  // save the result to be returned
+            setNextLocalResult();            // calculate and store the next result
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in Object Lookup", getMetadata());
+    }
+
+    @Override
+    protected void openLocal(DynamicContext context) {
+        this._currentDynamicContext = context;
+
+        _child.open(context);
+        setNextLocalResult();
+    }
+
+    private void setNextLocalResult() {
+        _nextLocalResult = null;
+        if (_expression.hasNext()) {
+            _nextLocalResult = _expression.next();
+        } else {
+            _expression.close();
+
+            if (_child.hasNext()) {
+                FlworTuple tuple = _child.next();
+                DynamicContext dynamicContext = new DynamicContext(tuple);
+                _expression.open(dynamicContext);
+                _nextLocalResult = _expression.next();
+            } else {
+                _child.close();
+                this._hasNext = false;
+            }
+        }
+
+        if (_nextLocalResult != null) {
```

Comment:
I think it would be cleaner to set _hasNext to true always at the same time as _nextLocalResult is set, and end each branch with a return; statement to exit the method. This makes the code easier to read.

Reply:
I was stuck between that choice and the current implementation :) I'll make the change right away